torch/fx/passes/_tensorify_python_scalars.py

```diff
@@ -244,13 +260,72 @@ def _sympy_interp(expr: sympy.Expr) -> MetaProxy:
                         )
 
                     node.replace_all_uses_with(replacement_proxy.node)
-
                     graph.erase_node(node)
+                    deleted.add(node)
 
     # DCE symbols (which are guaranteed to be pure) only
-    for proxy in reversed(expr_to_sym_proxy.values()):
-        if len(proxy.node.users) == 0 and proxy.node.op != "placeholder":
-            graph.erase_node(proxy.node)
+    dce(graph, deleted)
+
+    # Now do one more pass that specializes all symfloats we didn't manage
+    # to tensorify away.
```

Comment:
I think you can write this pass differently, that avoids the need to do another DCE at the end. The idea is to do the pass is REVERSE, rather than front-to-back.

When doing the pass in reverse, we look for operators that have backed SymFloat arguments, where we have a hint for it (maybe we also want to check if the operator supported it? Not sure, my primary fear is that we are going to obliterate something like math.floor(math.sqrt(x.size(0)), I want these to stay in the graph and not specialize them. Maybe a better test is something like, check that there aren't any free regular integer symbols in the expression). When this occurs, we issue a guard (you also have a choice here, whether or to guard on the expression, or guard on each of the free unbacked symfloats individually; the latter is less risky but the former preserves more dynamism.) And then you can immediately swap the node.

And then if you like, you can also DCE as you go, because if you kill all the users of a symfloat expression then things will start become eliminatable as you go backwards.

Will be good to write some tests for this.

Does this make sense?

Reply:
Can you elaborate on this case?

```
math.floor(math.sqrt(x.size(0))
```

`x.size(0)` returns a SymInt not a SymFloat right? We only specialize away SymFloats so this shouldn't be a concern?

Regarding writing the pass in reverse to eliminate a DCE pass, that does make sense. Given how hard these diffs have been to get green and this pass technically does work correctly, I'll try and stack a refactor diff on top of this one that does the pass in reverse.

Reply:
I'm talking about in my hypothetical rewrite. you will process the output of math.floor first, which potentially is eligible for specialization.

Reply:
I don't mind you stacking on top but if we're not in a rush I will just wait to land all of them together

Reply:
The only benefit of landing these earlier is we can start kicking the tires and seeing what breaks in the wild. I don't have strong opinions either way but generally I like to err on the side of merging these semantically correct changes earlier and then landing subsequent refactors. The time delay is also useful to discern if breakages are actually due to the refactor itself.

Reply:
This is a reasonable argument, but my experience in pytorch has taught me that landing closely interrelated changes at too close of an interval (a few days, which is the case here since you got CI working on all of them) together is more pain than gain. The problem is that if the early PR *does* cause a problem not only does that PR have to be backed out, but all of the ones that depend on it on top. Sure you did find out that it was the original PR and not the refactor that caused the problem... but if there's a minimal repro you would have easily been able to figure that out anyway.

I'm sympathetic to keeping the refactor PR stacked on top for CI signal (because maybe you did break something in the refactor that was fine in the original), and I'm also sympathetic to going ahead and landing the original PR if it's all ready to go. But you don't even have ciflow/trunk on this PR (so you can't land it right now), AND I see you are clean CI on the two PRs stacked on top, as of Sunday evening. Might as well squash them together.

Reply:
Sounds good. I think you were one step ahead of me. FWIW I didn't anticipate being able to get the CI green so quickly :p.